lib/mayaUsd/render/MaterialXGenOgsXml/ShaderGenUtil.cpp

```diff
@@ -0,0 +1,383 @@
+#include "ShaderGenUtil.h"
+
+#include <mayaUsd/render/MaterialXGenOgsXml/CombinedMaterialXVersion.h>
+
+#include <MaterialXCore/Document.h>
+
+#include <sstream>
+
+namespace MaterialXMaya {
+namespace ShaderGenUtil {
+
+namespace {
+const std::string SURFACEMATERIAL_CATEGORY("surfacematerial");
+const std::string SURFACESHADER_TYPE("surfaceshader");
+
+const std::set<std::string> _mtlxTopoNodeSet = {
+    // Topo affecting nodes due to object/model/world space parameter
+    "position",
+    "normal",
+    "tangent",
+    "bitangent",
+    // Topo affecting nodes due to channel index.
+    "texcoord",
+    // Color at vertices also affect topo, but we have not locked a naming scheme to go from index
+    // based to name based as we did for UV sets. We will mark them as topo-affecting, but there is
+    // nothing we can do to link them correctly to a primvar without specifying a naming scheme.
+    "geomcolor",
+    // Geompropvalue are the best way to reference a primvar by name. The primvar name is
+    // topo-affecting. Note that boolean and string are not supported by the GLSL codegen.
+    "geompropvalue",
+    // Swizzles are inlined into the codegen and affect topology.
+    "swizzle",
+    // Conversion nodes:
+    "convert",
+    // Constants: they get inlined in the source.
+    "constant",
+#if MX_COMBINED_VERSION < 13808
+    // Switch, unless all inputs are connected. Bug was fixed in 1.38.8.
+    "switch",
+#endif
+#if MX_COMBINED_VERSION == 13807
+    // Dot became topological in 1.38.7. Reverted in 1.38.8.
+    // Still topological for filename though.
+    "dot",
+#endif
+};
+} // namespace
+
+const std::string& TopoNeutralGraph::getMaterialName()
+{
+    // A material node is always the first node created and will be named N0
+    static const std::string kMaterialName("N0");
+    return kMaterialName;
+}
+
+TopoNeutralGraph::TopoNeutralGraph(const mx::ElementPtr& material)
+{
+    if (!material) {
```

Comment:
Having a look at the test cases really helps understand what this code is doing.